src/loan.rs

```diff
@@ -54,18 +58,44 @@ static COVENANT_PK: &str = "03b9b6059008e3576aad58e05a3a3e37133b05f68cda8535ec09
 /// identify themself by providing a signature on `borrower_pk`. To
 /// ensure that the borrower does indeed repay the loan, the script
 /// will check that the spending transaction has the
-/// `repayment_output` as vout 0.
+/// `repayment_principal_output` as vout 0.
 ///
-/// The first element of the covenant descriptor is the shared
-/// `covenant_pk`, which is only used to verify that the transaction
-/// data on the witness stack matches the transaction which triggered
-/// the call.
+/// Additionally, the contract can be spent by providing a signature
+/// (and the corresponding message) from a valid oracle proving that
+/// the price of the collateral has dropped below a threshold
+/// determined at contract creation time.
 #[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
-struct CollateralContract {
+pub struct CollateralContract {
+    /// Only includes the two branches that can be expressed using miniscript:
+    ///
+    /// - Liquidation after expiration.
+    ///
+    /// - Collateral reclamation after repayment.
+    ///
+    /// Kept around to more easily construct the witness stack for
+    /// those two branches.
     descriptor: Descriptor<PublicKey>,
+    /// The actual script, including the dynamic liquidation branch.
+    raw_script: Script,
+    /// The script after `OP_CODESEPARATOR`.
+    ///
+    /// This is the only part of the script which is included in the
+    /// transaction sighash which is signed with `COVENANT_SK`.
+    ///
+    /// In a regular `CovenantDescriptor` this is just
+    /// `OP_CHECKSIGVERIFY` and `OP_CHECKSIGFROMSTACK`. In our case,
+    /// it's `OP_CHECKSIGVERIFY`, `OP_CHECKSIGFROMSTACK` and
+    /// `OP_ENDIF`.
+    cov_script: Script,
```

Comment:
This is always constant yeah? Can we model it as a module-level constant then?

Reply:
Good point!

Reply:
I almost got it working but the borrow checker blocked me. Check it out if you want: https://github.com/comit-network/baru/commit/d9ac0f26b8a5a27461346588d69b96507685e3fe.

Reply:
Solved the problem I mentioned above although it's not ideal: https://github.com/comit-network/baru/pull/25/commits/ed0c50462a77027f795f2f24578aaa2dc7a38394.